add a GotoLineModal

jfo opened the pull request:
Associated Issue: #4210 

### Summary of Changes

This commit adds a gotoline modal/functionality triggered with `:` in
the main debugger source window.

- [x] `Cmd+:` opens the modal
- [x] Entering a line number takes you to that line in the current source
- [x] Clicking "x" closes the panel
- [x] unusable input (i.e. letters) result in a noop

![goto](https://user-images.githubusercontent.com/4204764/31310324-68610ff6-ab96-11e7-9485-242933fde281.gif)

This is not ready to merge but it is ready for feedback. I've commented on the diff with some things! Also needs tests, would like some guidance on that via [SymbolModal.spec.js](https://github.com/devtools-html/debugger.html/blob/master/src/components/tests/SymbolModal.spec.js), I assume.

jfo commented on assets/panel/debugger.properties:
```diff
@@ -648,6 +648,14 @@ watchExpressionsSeparatorLabel2=\u0020→
+# LOCALIZATION NOTE(symbolSearch.search.key2): The Key Shortcut for
+# searching for a function or variable
+gotoLineModal.key=:
+
```
idk if this is the right place for this.

jfo commented on src/components/App.js:
```diff
@@ -47,6 +47,8 @@ import EditorTabs from "./Editor/Tabs";
 
 import SymbolModal from "./SymbolModal";
 
+import GotoLineModal from "./GotoLineModal";
```
this whole pattern is from @codehag's diff on the associated issue. 

It feels like these (SymbolModal, ShortcutModal, GotoLineModal) should all be the same thing, but I thought this would be a good start before merging them together.

jasonLaster replied:
Yeah... we do want to consolidate the symbol + gotoline modal and search for file.

+ `cmd+p`
+ `cmd+:`
+ `cmd+shift+o`

jfo replied:
- [cmd+p](https://github.com/devtools-html/debugger.html/blob/master/assets/panel/debugger.properties#L141)
- [cmd+shift+o](https://github.com/devtools-html/debugger.html/blob/master/assets/panel/debugger.properties#L153)

For those shortcuts... `CmdOrCtrl+Shift+;` works, but `CmdOrCtrl+:` does not work, nor does `CmdOrCtrl+Shift+:`. Is this correct? This could be revisted when you unify them, whatever you'd like for right now is fine by me :)

jfo commented on src/components/GotoLineModal.js:
```diff
@@ -0,0 +1,118 @@
+  query: ?string
+};
+
+import "./SymbolModal.css";
```
I'm pretty sure this file is actually not being used anywhere (it's also called in `SymbolModal.js` but the `.symbol-modal-wrapper` and `.symbol-modal` classes are never applied. I might suggest merging the [remaining rules](https://github.com/devtools-html/debugger.html/blob/master/src/components/SymbolModal.css#L18-L31) into `shared/Modal.css` instead but that's just outside the scope of this pr.

wldcordeiro replied:
It looks the those rules are applied?

https://github.com/devtools-html/debugger.html/blob/master/src/components/SymbolModal.js#L227 `.input-wrapper`

and then the result list? https://github.com/devtools-html/debugger.html/blob/master/src/components/SymbolModal.js#L213

jfo replied:
Sorry, I wasn't clear, I meant that only `.symbol-modal-wrapper` and `.symbol-modal` classes don't seem to be applied and the remaining rules that I linked to might be incorporated into `Modal.css` in the process of merging the modals together more, but I'm not planning to do that on this PR.

flodolo commented on assets/panel/debugger.properties:
```diff
@@ -648,6 +648,14 @@ watchExpressionsSeparatorLabel2=\u0020→
 
+# LOCALIZATION NOTE(gotoLineModal.placeholder): The placeholder
+# text displayed when the user searches for specific lines in a file
+gotoLineModal.placeholder=Go to line...
```
Use …, not ...

jfo replied:
It took me a while to figure out what you meant by this, that the first is a unicode character is hard to discern from the text. Putting them in `pre` fences would help distinguish them i.e.
 `...` vs `…`.

flodolo replied:
Good point, I'll keep it in mind.

(in case you're seeing this comment twice, I re-added comment in the right thread).

flodolo commented on assets/panel/debugger.properties:
```diff
@@ -648,6 +648,14 @@ watchExpressionsSeparatorLabel2=\u0020→
+# text displayed when the user searches for specific lines in a file
+gotoLineModal.placeholder=Go to line...
+
+# LOCALIZATION NOTE(symbolSearch.search.key2): The Key Shortcut for
```
Please stop copying localization comments without updating them

codehag replied:
A more constructive way to say this would be "Please update this localization note to reflect the new key"

flodolo replied:
It wasn't meant as an hostile comment towards the author, if it came across like that, I apologize.

For context: this is happening in several PRs every week, so I think there's a general issue to solve. Commenting strings is not mandatory, comments are there to give context to translators. On the other hand, I have the feeling most developers think they are, but copying and paste an existing comment is not the solution.

jfo replied:
It did come across like that, so thank you for apologizing. I have no knowledge of what other PR's look like (this is my very first one of any substance) or that the comments are optional, I'm just trying to match what's around the usage with the expectation of being corrected if necessary, so thanks for catching it, it was an oversight.

flodolo commented on assets/panel/debugger.properties:
```diff
@@ -648,6 +648,14 @@ watchExpressionsSeparatorLabel2=\u0020→
+
+# LOCALIZATION NOTE(symbolSearch.search.key2): The Key Shortcut for
+# searching for a function or variable
+gotoLineModal.key=:
```
Does this even work? IIRC you need to press shift to get to that keyboard in English, and that's a modifier.

jfo replied:
It does work. There is an attached gif.

codehag replied:
Hi, thanks for helping our contributors with their prs. I would like to point out that the wording of this comment can be read as aggressive, and we want our contributors to feel safe and that their changes are welcome. Phrasing such as "Does this even work" can be off-putting for a contributor, and can also turn away potential contributors who are looking through our PRs to get a sense of the community. It is especially problematic to have this on something that is clearly labelled as a work in progress, since the person may be looking for feedback.

If you find something amiss, think of how to word things in a way that are not judgemental or polarizing.

jfo replied:
Sorry. Yes. You do have to press shift to get to that symbol, but this syntax works to trigger that. In any case, it's a temporary key that can be changed to anything that makes the most sense.

flodolo replied:
As above, this comment wasn't meant as hostile or derogatory. 

I can't know in which state a PR is when it's open (I'm not a developer, I work on localization), if it's testable or just an early WIP (it looked like that), hence the question "Does : this even work?". There was no underlying implication, i.e. "are you even testing your code?", in my question.

I don't know exactly how devtools or debugger manage shortcuts internally, but I don't think it would work for a feature in Firefox, since we stumbled upon it for shortcuts on special characters like [ or ]

codehag replied:
thanks for clarifying! And again thanks for helping :)

flodolo commented on assets/panel/debugger.properties:
```diff
@@ -648,6 +648,14 @@ watchExpressionsSeparatorLabel2=\u0020→
+# text displayed when the user searches for specific lines in a file
+gotoLineModal.placeholder=Go to line…
+
+# LOCALIZATION NOTE(gotoLineModal.key): The Key Shortcut for
```
You still need to update the comment itself (you only updated the reference to the string ID, but the text describes a different function).

At this point I would suggest to drop this comment completely, put this string near the one above (that already has a comment explaining the function), and call this string `gotoLineModal.commandkey` to avoid any potential confusion. 

There's no rule set in stone for key naming, but `.key` is more prone to misinterpretation that `.commandkey`.

jfo replied:
Oh that's a good idea.

To reiterate, `Cmd+:` is definitely temporary, just needed something to use for testing b/c the suggested (in the linked issue) `Cmd+g` is already being used. I think ultimately just `:` would be a good shortcut _from an already opened modal_ but definitely not as a global shortcut key.

flodolo replied:
Good point, I'll keep it in mind.

jasonLaster commented on src/components/GotoLineModal.js:
```diff
@@ -0,0 +1,118 @@
+      this.closeModal();
+    } else if (e.key === "Tab") {
+      this.closeModal();
+    }
```
I think we might want to handle "Escape" too

jfo replied:
Escape already closes out the modal in both SymbolModal and GotoLineModal, it's handled more globally [here](https://github.com/devtools-html/debugger.html/blob/master/src/actions/ui.js#L7) via [here](https://github.com/devtools-html/debugger.html/blob/master/src/components/App.js#L118) and [here](https://github.com/devtools-html/debugger.html/blob/master/src/components/App.js#L105).

jasonLaster commented on src/components/GotoLineModal.js:
```diff
@@ -0,0 +1,118 @@
+        selectSource(selectedSource.get("id"), { line: linenumber });
+      }
+      this.closeModal();
+    } else if (e.key === "Tab") {
```
this is just a style thing, but i tend to prefer early returns

```js
if (e.key === "Enter" && query != null) {
 if (!isNaN(linenumber)) {
   selectSource(selectedSource.get("id"), { line: linenumber })
 }
 return this.closeModal();
}

if (e.key === "Tab") {
  return this.closeModal();
}
```

jfo replied:
I'll defer to you on that 😃 ! But, as those return values aren't being dispatched on in the caller, I'd prefer 

```js
if (e.key === "Enter" && query != null) {
 if (!isNaN(linenumber)) {
   selectSource(selectedSource.get("id"), { line: linenumber })
 }
 this.closeModal();
 return;
}

if (e.key === "Tab") {
  this.closeModal();
  return;
}
```

When I was moving some logic to this from `SymbolModal.js` I had trouble grokking [this line](https://github.com/devtools-html/debugger.html/blob/master/src/components/SymbolModal.js#L109), because I assumed the `return` value was significant. It can't be, since [updateResults](https://github.com/devtools-html/debugger.html/blob/master/src/components/SymbolModal.js#L131-L145) always returns undefined, but semantically that was unclear.

Is there a technical reason to avoid `else if` in favor of multiple `if` blocks? Like a branching implementation thing?

jasonLaster commented on assets/panel/debugger.properties:
```diff
@@ -648,6 +648,14 @@ watchExpressionsSeparatorLabel2=\u0020→
+
+# LOCALIZATION NOTE(gotoLineModal.key): The Key Shortcut for
+# searching for a specific line
+gotoLineModal.key=:
```
I'm surprised that this is not `cmd+:`

jasonLaster commented on src/components/App.js:
```diff
@@ -150,6 +161,28 @@ class App extends Component {
+      return closeActiveSearch();
+    }
+
+    setActiveSearch("line");
```
we'll likely need to update the flow definition for the possible search types in `reducers/ui.js`

jfo replied:
do you mean [here](https://github.com/devtools-html/debugger.html/blob/master/src/reducers/ui.js#L21)? I am surprised flow did not catch that for me in a warning... I guess I don't really know how flow works 😃

jasonLaster commented:
by the way, we have an open PR to fix the position of the modal: https://github.com/devtools-html/debugger.html/pull/4277

it should always be pinned to the top...

jfo commented:
Thanks all for the good feedback! Will this require tests?

jasonLaster commented:
@jfo i don't think so

jfo commented:
Hi all, I've addressed the outstanding reviews and rebased against master, I think this should be ready to merge now but let me know if there is anything else needed.

Thanks a lot! 🙌

jasonLaster commented:
🎁
